src/components/Attachment/hooks/useAudioController.ts

```diff
@@ -1,17 +1,61 @@
 import { useCallback, useEffect, useRef, useState } from 'react';
+import { useChannelActionContext, useTranslationContext } from '../../../context';
+
+export const elementIsPlaying = (audioElement: HTMLAudioElement | null) =>
+  audioElement && !(audioElement.paused || audioElement.ended);
 
 export const PROGRESS_UPDATE_INTERVAL = 100;
 
-export const useAudioController = () => {
+const DEFAULT_PLAYBACK_RATES = [1.0, 1.5, 2.0];
+
+type AudioControllerParams = {
+  /** Audio duration in seconds. */
+  durationSeconds?: number;
+  /** An array of fractional numeric values of playback speed to override the defaults (1.0, 1.5, 2.0) */
+  playbackRates?: number[];
+};
+
+export const useAudioController = (params: AudioControllerParams = {}) => {
+  const { addNotification } = useChannelActionContext('useAudioController');
+  const { t } = useTranslationContext('useAudioController');
+  const { durationSeconds } = params;
+  const playbackRates = params.playbackRates || DEFAULT_PLAYBACK_RATES;
   const [isPlaying, setIsPlaying] = useState(false);
   const [progress, setProgress] = useState(0);
+  const [secondsElapsed, setSecondsElapsed] = useState(durationSeconds);
+  const [playbackRateIndex, setPlaybackRateIndex] = useState<number>(0);
 
   const audioRef = useRef<HTMLAudioElement | null>(null);
+  const playInterval = useRef<number>();
 
   const togglePlay = useCallback(() => {
-    setIsPlaying((playing) => !playing);
+    if (!audioRef.current) return;
+    clearInterval(playInterval.current);
+    if (!elementIsPlaying(audioRef.current)) {
+      audioRef.current.play();
+
+      playInterval.current = window.setInterval(() => {
+        if (!audioRef.current) return;
+        const { currentTime, duration } = audioRef.current;
+        setProgress((currentTime / duration) * 100);
+        setSecondsElapsed(currentTime);
+      }, PROGRESS_UPDATE_INTERVAL);
+      setIsPlaying(true);
+    } else {
+      audioRef.current.pause();
+      setIsPlaying(false);
+    }
   }, []);
 
+  const increasePlaybackRate = useCallback(() => {
+    setPlaybackRateIndex((prev) => {
+      if (!audioRef.current) return prev;
+      const nextIndex = prev === playbackRates.length - 1 ? 0 : prev + 1;
+      audioRef.current.playbackRate = playbackRates[nextIndex];
+      return nextIndex;
+    });
+  }, [playbackRates]);
```

Comment:
If someone passes inline playbackRates, this memoization will be busted. I'm not sure if it is worth it to have this logic wrapped with `useCallback`